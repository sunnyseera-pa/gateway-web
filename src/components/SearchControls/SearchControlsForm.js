import React from 'react';
import { useFormikContext } from 'formik';
import { Col, Row } from 'react-bootstrap';
import { useTranslation } from 'react-i18next';
import SearchInput from '../SearchInput';
import SortDropdown from '../SortDropdown';
import useCommonStyles from '../../hooks/useCommonStyles';

const SearchControlsForm = ({ sortProps, inputProps, type, mt, mb, mr, ml, width, minWidth, maxWidth }) => {
	const { t } = useTranslation();
	const commonStyles = useCommonStyles({ mt, mb, mr, ml, width, minWidth, maxWidth });

	const {
		submitForm,
		values: { search, sortBy, sortDirection },
		values,
		setFieldValue,
	} = useFormikContext();

	const handleKeyDown = React.useCallback(e => {
		if (e.key === 'Enter') {
			submitForm();
		}
	}, []);

	const handleReset = React.useCallback(() => {
		setFieldValue('search', '');
		submitForm();
	}, [values]);

	const handleChange = React.useCallback(({ target: { name, value } }) => {
		setFieldValue(name, value);

		const { onChange } = inputProps;

		if (onChange) onChange(value);
	}, []);

	const handleOnSort = React.useCallback(
		criteria => {
			const { value, direction } = criteria;
			const { onSort } = sortProps;

			setFieldValue('sortBy', value);
			setFieldValue('sortDirection', direction);

			if (onSort) onSort(criteria);

			submitForm();
		},
		[values]
	);

	return (
		<Row className={commonStyles}>
			<Col lg={6}>
				<SearchInput
					placeholder={t('search.placeholder', { type: type.toLowerCase() })}
					variant='primary'
					{...inputProps}
					value={search}
					onKeyDown={handleKeyDown}
					onChange={handleChange}
					onReset={handleReset}
				/>
			</Col>
			<Col lg={2} />
			{sortProps && (
				<Col lg={4} className='d-flex justify-content-end'>
					<SortDropdown {...sortProps} onSort={handleOnSort} value={sortBy} direction={sortDirection} />
				</Col>
			)}
		</Row>
	);
};

export default SearchControlsForm;
