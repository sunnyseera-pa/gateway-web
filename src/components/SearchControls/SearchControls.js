import React from 'react';
import { Formik } from 'formik';
import SearchControlsForm from './SearchControlsForm';

const SearchControls = ({ onSubmit, isLoading, sortProps, inputProps, ...outerProps }) => {
	if (isLoading) return null;

	return (
		<Formik initialValues={{ search: inputProps ? inputProps.value : '', sortBy: sortProps ? sortProps.value : '' }} onSubmit={onSubmit}>
			<SearchControlsForm sortProps={sortProps} inputProps={inputProps} {...outerProps} />
		</Formik>
	);
};

export default SearchControls;
