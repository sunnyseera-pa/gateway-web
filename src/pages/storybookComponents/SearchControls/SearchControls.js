import React from 'react';
import { Formik } from 'formik';
import SearchControlsForm from './SearchControlsForm';

const SearchControls = ({ onSubmit, isLoading, sortProps, ...outerProps }) => {
	if (isLoading) return null;

	return (
		<Formik initialValues={{ search: '', sortBy: sortProps ? sortProps.value : '' }} onSubmit={onSubmit}>
			<SearchControlsForm sortProps={sortProps} {...outerProps} />
		</Formik>
	);
};

export default SearchControls;
