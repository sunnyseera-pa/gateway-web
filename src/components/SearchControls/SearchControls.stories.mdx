import { Meta, Story, Canvas } from '@storybook/addon-docs';

import SearchControls from './SearchControls';

<Meta
	title='Core/SearchControls'
	component={SearchControls}
	args={{
		type: 'dataset',
		sortProps: { options: ['relevance', 'recentlyadded'], value: 'relevance' },
		inputProps: {
			value: 'covid 19',
		},
		width: '900px',
	}}
/>

<Canvas>
	<Story name='Basic'>{props => <SearchControls {...props} />}</Story>
</Canvas>
