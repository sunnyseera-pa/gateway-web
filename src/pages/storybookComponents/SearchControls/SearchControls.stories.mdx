import { Meta, Story, Canvas } from '@storybook/addon-docs';

import SearchControls from './SearchControls';

<Meta title='Core/SearchControls' component={SearchControls} />

<Canvas>
	<Story name='Basic'>{props => <SearchControls {...props} />}</Story>
</Canvas>
