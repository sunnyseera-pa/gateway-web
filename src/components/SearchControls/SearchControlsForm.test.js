import { render, waitFor } from '@testing-library/react';
import { act } from '@testing-library/react-hooks';
import React from 'react';
import SearchControls from '.';

const mockOnSort = jest.fn();
const mockOnChange = jest.fn();
const mockOnSubmit = jest.fn();
const mockOnReset = jest.fn();

const props = {
	onSubmit: mockOnSubmit,
	isLoading: true,
	sortProps: {
		onSort: mockOnSort,
		value: 'relevance',
		options: ['relevance', 'recentlyadded'],
	},
	inputProps: {
		onReset: mockOnReset,
		onChange: mockOnChange,
		value: 'dataset',
	},
};

let wrapper;

describe('Given the SearchControls component', () => {
	describe('When it is rendered', () => {
		beforeAll(() => {
			wrapper = render(<SearchControls {...props} />, {
				wrapper: Providers,
			});
		});

		it('Then matches the previous snapshot', () => {
			expect(wrapper.container.textContent).toBeFalsy();
		});

		describe('And it is not loading', () => {
			beforeAll(() => {
				wrapper = render(<SearchControls {...props} isLoading={false} />, {
					wrapper: Providers,
				});
			});

			it('Then matches the previous snapshot', () => {
				expect(wrapper.container).toMatchSnapshot();
			});

			describe('And the input is changed', () => {
				beforeAll(() => {
					const input = wrapper.container.querySelector('input');

					fireEvent.change(input, { target: { name: 'search', value: 'collection' } });
				});

				it('Then calls onChange', async () => {
					await waitFor(() => expect(mockOnChange).toHaveBeenCalledWith('collection'));
				});

				describe('And the input is entered on', () => {
					beforeAll(() => {
						const input = wrapper.container.querySelector('input');

						fireEvent.keyDown(input, { key: 'Enter' });
					});

					it('Then calls onSubmit', async () => {
						await waitFor(() =>
							expect(mockOnSubmit.mock.calls[0][0]).toMatchObject({
								search: 'collection',
								sortBy: 'relevance',
							})
						);
					});
				});
			});

			describe('And the sort is changed on', () => {
				beforeAll(() => {
					const button = wrapper.container.querySelector('.ui-Dropdown button');

					fireEvent.click(button);

					const link = wrapper.container.querySelectorAll('a')[1];

					fireEvent.click(link);
				});

				it('Then calls onSort', async () => {
					await waitFor(() => expect(mockOnSort.mock.calls[0][0]).toEqual('recentlyadded'));
				});

				it('Then calls onSubmit', async () => {
					await waitFor(() => expect(mockOnSubmit).toHaveBeenCalled());
				});
			});
		});
	});
});
