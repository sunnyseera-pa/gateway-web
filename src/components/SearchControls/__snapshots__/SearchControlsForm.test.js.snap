// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the SearchControls component When it is rendered And it is not loading Then matches the previous snapshot 1`] = `
<div>
  <div
    class="css-1sivg29 row"
  >
    <div
      class="col-lg-6"
    >
      <div
        class="css-koyxv1-Input form-group"
      >
        <label
          class="form-label"
        />
        <div
          class="ui-Input ui-SearchInput css-1sivg29 css-1r1cho1-Input input-group"
        >
          <div
            class="css-haly75-Input input-group-prepend"
          >
            <button
              fill="green700"
              name="search"
            >
              Icon
            </button>
          </div>
          <input
            class="form-control"
            name="search"
            placeholder="search.placeholder"
            value="dataset"
          />
          <div
            class="css-x0rxt-Input input-group-append"
          >
            <button
              color="purple500"
              name="clear"
              role="button"
            >
              Icon
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="col-lg-2"
    />
    <div
      class="d-flex justify-content-end col-lg-4"
    >
      <div
        class="ui-SortDropdown css-1sivg29 css-16k9oib-SortDropdown"
      >
        <div
          class="ui-Dropdown css-1sivg29 css-1quv89l-SortDropdown dropdown"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle btn btn-primary"
            type="button"
          >
            sortby.options.relevance
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
